library/kani/src/aliasing.rs

```diff
@@ -0,0 +1,316 @@
+    impl Permission {
+        const UNIQUE: u8 = 0;
+        const SHAREDRW: u8 = 1;
+        const SHAREDRO: u8 = 2;
+        const DISABLED: u8 = 3;
+
+        fn grants(access: bool, tag: u8) -> bool {
+            tag != Self::DISABLED && (access != Access::WRITE || tag != Self::SHAREDRO)
+        }
+    }
+
+    /// Associate every pointer object with a permission
+    static mut PERMS: ShadowMem<u8> = ShadowMem::new(0);
+
+    pub(super) mod monitors {
+        static mut STATE: bool = false;
+        static mut OBJECT: usize = 0;
+        static mut OFFSET: usize = 0;
+        static mut STACK_TAGS: [u8; STACK_DEPTH] = [0; STACK_DEPTH];
+        static mut STACK_PERMS: [u8; STACK_DEPTH] = [0; STACK_DEPTH];
+        static mut STACK_TOP: usize = 0;
+
+        #[non_exhaustive]
+        struct MonitorState;
+        impl MonitorState {
+            const UNINIT: bool = false;
+            const INIT: bool = true;
+        }
+
+        use super::*;
+
+        /// Monitors:
+        /// If there are K bytes in the address space,
+        /// every stacked borrows instrumentation has
+        /// between 0 and K monitors.
+        /// These monitors track a single byte of the program,
+        /// associating it with a stack of pointer values
+        /// (represented by tags).
+        /// Whenever a pointer borrows an object containing
+        /// the byte, its tag is pushed to the stack;
+        /// when a read or write is performed through this pointer,
+        /// writes from pointers above its location on the stack
+        /// are disabled.
+        /// This function prepares N monitors,
+        /// writes them to global heap memory, then
+        /// stores them in pointers.
+        /// An N+1th monitor is allocated as a "garbage"
+        /// area to be used when no monitor is picked.
+        pub fn prepare_monitors() {
+            unsafe {
+                OBJECT = 0usize;
+                OFFSET = 0usize;
+                STATE = MonitorState::UNINIT;
+                STACK_TAGS = [NEXT_TAG; STACK_DEPTH];
+                STACK_PERMS = [Permission::UNIQUE; STACK_DEPTH];
+                STACK_TOP = 0usize;
+            }
+        }
+
+        /// Initialize local when track local is true, picking a monitor,
+        /// and setting its object and offset to within pointer.
+        pub(super) unsafe fn track_local<U>(tag: u8, pointer: *const U) {
+            // Decide whether to initialize the stacks
+            // for location:location+size_of(U).
+            // Offset has already been picked earlier.
+            unsafe {
+                if demonic_nondet() && STATE == MonitorState::UNINIT {
+                    STATE = MonitorState::INIT;
+                    OBJECT = pointer_object(pointer);
+                    OFFSET = 0;
+                    crate::assume(OFFSET < std::mem::size_of::<U>());
+                    STACK_TAGS[STACK_TOP] = tag;
+                    STACK_PERMS[STACK_TOP] = Permission::UNIQUE;
```

Comment:
That is not actually checked there. Rather, it is only possible to obtain a mutable pointer by first going through an immutable reference. The mutable pointer will then not have write perms.

This can be see in the original SB calculus in this example here:

https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=b82326f44255c2e10eac00a4c82cb5e4

Reply:
(If you run with miri)